File: 백준/Gold/9663. N－Queen/N－Queen.cc
#include<iostream>

using namespace std;

int n;

bool vcol[15] = { false, };
bool vrd[30] = { false, };
bool vld[31] = { false, };
int ans = 0;

void bt(int idx) {
	if (idx >= n) {
		ans++;
		return;
	}
	for (int i = 0; i < n; ++i) {
		if (vcol[i] || vld[(idx-i)+15] || vrd[idx+i]) continue;
		else {
			vcol[i] = true;
			vld[idx - i + 15] = true;
			vrd[idx + i]=true;
			bt(idx + 1);
			vcol[i] = false;
			vld[idx - i+15] = false;
			vrd[idx + i] = false;
		}
	}
	return;
}

int main() {
	cin >> n;
	bt(0);
	cout << ans;
}

File: 백준/Gold/9663. N－Queen/N－Queen.py


def bt(idx) :
    if (idx >= n) :
        global ans
        ans=ans+1
        return
    for i in range(0,n):
        if (vcol[i] or vld[(idx-i)+15] or vrd[idx+i]) : continue
        else : 
            vcol[i] = True
            vld[idx - i + 15] = True
            vrd[idx + i]=True
            bt(idx + 1)
            vcol[i] = False
            vld[idx - i+15] = False
            vrd[idx + i] = False


n=int(input());

vcol = [False]*16
vrd = [False]*31
vld = [False]*31
ans = 0

bt(0)
print(ans)